scripts/common/common.sh

```diff
@@ -1151,3 +1156,274 @@ function cmd_retry() {
+    [ "$(common_upgrade_compare_versions "$current_version" "$from_version")" = "1" ] && \
+    [ "$(common_upgrade_compare_versions "$current_version" "$to_version")" != "1" ]
+}
+
+# common_upgrade_max_version will return the greater of the two versions.
+function common_upgrade_max_version() {
+    local a="$1"
+    local b="$2"
+    if [ "$(common_upgrade_compare_versions "$a" "$b")" = "1" ]; then
+        echo "$a"
+    else
+        echo "$b"
+    fi
+}
+
+# common_upgrade_print_list_of_minor_upgrades prints message of minor versions that will be
+# upgraded. e.g. "1.0.x to 1.1, 1.1 to 1.2, 1.2 to 1.3, and 1.3 to 1.4"
+function common_upgrade_print_list_of_minor_upgrades() {
+    local from_version="$1"
+    local to_version="$2"
+
+    printf "This involves upgrading from "
+    local first_minor=
+    local last_minor=
+    first_minor=$(common_upgrade_major_minor_to_minor "$from_version")
+    last_minor=$(common_upgrade_major_minor_to_minor "$to_version")
+
+    local minor=
+    for (( minor=first_minor ; minor<last_minor ; minor++ )); do
+        if [ "$minor" -gt "$first_minor" ]; then
+            printf ", "
+            if [ "$((minor + 1))" -eq "$last_minor" ]; then
+                printf "and "
+            fi
+            printf "1.%s to 1.%s" "$minor" "$((minor + 1))"
+        else
+            printf "1.%s.x to 1.%s" "$minor" "$((minor + 1))"
+        fi
+    done
+    printf ".\n"
+}
+
+# common_upgrade_major_minor_to_major returns the major version of a major.minor version.
+function common_upgrade_major_minor_to_major() {
+    echo "$1" | cut -d. -f1
+}
+
+# common_upgrade_major_minor_to_minor returns the minor version of a major.minor version.
+function common_upgrade_major_minor_to_minor() {
+    echo "$1" | cut -d. -f2
+}
+
+# common_upgrade_version_to_major_minor returns the major.minor version of a semver version.
+function common_upgrade_version_to_major_minor() {
+    echo "$1" | cut -d. -f1,2
+}
+
+# common_list_images_in_manifest_file will list images in the given manifest file.
+function common_list_images_in_manifest_file() {
+    local manifest_file="$1"
+
+    local image_list=
+    for image in $(grep "^image " "$manifest_file" | awk '{print $3}' | tr '\n' ' ') ; do
+        image_list=$image_list" $(canonical_image_name "$image")"
+    done
+    echo "$image_list" | xargs # trim whitespace
+}
+
+# common_upgrade_merge_images_list will merge each list of images from the arguments into a single
+# list and deduplicate the list.
+function common_upgrade_merge_images_list() {
+    local images_list=
+    while [ "$1" != "" ]; do
+        images_list="$images_list $1"
+        shift
+    done
+    echo "$images_list" | tr " " "\n" | sort | uniq | tr "\n" " " | xargs
+}
+
+# common_upgrade_storage_check verifies that enough disk space exists based on the desired size.
+function common_upgrade_storage_check() {
```

Comment:
@laverya can this function be generic? can i pass in a factor or something if it cannot? i am hoping hot to duplicate the code.

Reply:
I think this should work no matter the archive (ie for both Rook and k8s upgrades) but I suspect the multipliers might be overly generous

Reply:
archive looks about the same size as it does extracted. i suppose that is how you arrived at the 2x. now i just need to figure out the 3.5x.

```
ethan@ethanm-k8s-upgrade-1:~/tmp$ du -d0 -h packages/
998M	packages/
ethan@ethanm-k8s-upgrade-1:~/tmp$ ls -lh kubernetes-1.26.3.tar.gz
-rw-rw-r-- 1 ethan ethan 935M Apr 14 18:21 kubernetes-1.26.3.tar.gz
```

Reply:
looks like images are a minimal part of the kubernetes add-on package but do extract to about 3.5 times their size

```
$ du -d1 -h
211M	./archives
702M	./extracted
$ tree -L 3
.
├── archives
│   ├── coredns.tar.gz
│   ├── etcd.tar.gz
│   ├── kube-apiserver.tar.gz
│   ├── kube-controller-manager.tar.gz
│   ├── kube-proxy.tar.gz
│   ├── kube-scheduler.tar.gz
│   └── pause.tar.gz
└── extracted
    ├── coredns
    │   ├── 49fecb8a03ea826d42ea1a538daadb2715c288ea07074b471fa2a20c0520a694
    │   ├── 5185b96f0becf59032b8e3646e99f84d9655dff3ac9e2605e0dc77f9c441ae4a.json
    │   ├── d6da605550dee544ebcd7ca201b46bfa28f433797739f92e13e53e9ea90ee93d
    │   ├── manifest.json
    │   └── repositories
    ├── etcd
    │   ├── 331eda04ad93d94b8339faf570c7e305fd7ce0c0178c17c0235ca68046a87f79
    │   ├── 4755f48758dddc0fdb92a5f566fcb824c442084bd40c49b013cbc48aeb6d05eb
    │   ├── 69aeb733f25c953881886383249e1b7bc1659da7c4e6a5f034522cbf9ab1b27b
    │   ├── 8e1863bdcad2d8bcd68b30876b0102c29e4006ac3a3a1b014264b5a2fa78d0b2
    │   ├── bb96165ed9b54fca8912aee1c2246909347de8441705f85ab5f210347cd44ac3
    │   ├── fce326961ae2d51a5f726883fd59d2a8c2ccc3e45d3bb859882db58e422e59e7.json
    │   ├── manifest.json
    │   └── repositories
    ├── kube-apiserver
    │   ├── 064abeed0b8606912527f2dea03299dcf800ebac6becafb1d8151953101fa28c
    │   ├── 1d9b3cbae03cea2a1766cfa5bf06a5a9c7a7bdbc6f5322756e29ac78e76f2708.json
    │   ├── 2a68f2e79aff7941693d95a523f0d3d7788b280136034b6cacfce41dcba0533d
    │   ├── b3cab702024f79855fe880df4e3790c135fb7e12e3ef19f899282802ad8a118d
    │   ├── manifest.json
    │   └── repositories
    ├── kube-controller-manager
    │   ├── 064abeed0b8606912527f2dea03299dcf800ebac6becafb1d8151953101fa28c
    │   ├── 90f633599691986a3169b4f42d6b783d11672d1daac6a1a6dadf65b80e3e3e47
    │   ├── b3cab702024f79855fe880df4e3790c135fb7e12e3ef19f899282802ad8a118d
    │   ├── ce8c2293ef09c9987773345638026f9f7aed16bc52e7a6ea507f0c655ab17161.json
    │   ├── manifest.json
    │   └── repositories
    ├── kube-proxy
    │   ├── 3f85f734ec2068a0deb9928d31c8478237851defe3d08f1f1f86b7223188e09e
    │   ├── 89a94e0e8b231e21813583d814ffe42e119dd0b6d1e90f2b72c40d4b8d893d42
    │   ├── 92ed2bec97a637010666d6c4aa4d69b672baec0fd5d236d142e4227a3a0557d8.json
    │   ├── manifest.json
    │   └── repositories
    ├── kube-scheduler
    │   ├── 064abeed0b8606912527f2dea03299dcf800ebac6becafb1d8151953101fa28c
    │   ├── 30de761707abea3223e7d3537045903e8f3dbe628781cb957c506cee7190f73a
    │   ├── 5a79047369329dff4a02e705e650664d2019e583b802416447a6a17e9debb62d.json
    │   ├── b3cab702024f79855fe880df4e3790c135fb7e12e3ef19f899282802ad8a118d
    │   ├── manifest.json
    │   └── repositories
    └── pause
        ├── 59d351ae471ad37a95dcd840c7d2b2931ed0f847bb889c4fbe4029e8cae60fd7
        ├── e6f1816883972d4be47bd48879a08919b96afcd344132622e4d444987919323c.json
        ├── manifest.json
        └── repositories
```

Reply:
break down of the bundle size

```
$ du -d1 -h .
97M	./assets
73M	./ubuntu-22.04
211M	./images
184M	./rhel-8
143M	./rhel-7
50M	./rhel-9
72M	./ubuntu-20.04
72M	./ubuntu-18.04
100M	./rhel-7-force
998M	.
```

Reply:
bundle

```
ethan@ethanm-k8s-upgrade-1:~/tmp$ ls -lh kubernetes-1.26.3.tar.gz
-rw-rw-r-- 1 ethan ethan 935M Apr 14 18:21 kubernetes-1.26.3.tar.gz
```

before

```
ethan@ethanm-k8s-upgrade-1:~$ sudo du -d0 /var/lib/containerd/
6204312	/var/lib/containerd/
ethan@ethanm-k8s-upgrade-1:~/tmp/packages/kubernetes/1.26.3$ sudo du -d0 -h /var/lib/containerd/
6.0G	/var/lib/containerd/
ethan@ethanm-k8s-upgrade-1:~$ sudo df /
Filesystem     1K-blocks     Used Available Use% Mounted on
/dev/root      203056560 21388628 181651548  11% /
ethan@ethanm-k8s-upgrade-1:~/tmp/packages/kubernetes/1.26.3$ sudo df -h /
Filesystem      Size  Used Avail Use% Mounted on
/dev/root       194G   21G  174G  11% /
```

after

```
ethan@ethanm-k8s-upgrade-1:~$ sudo du -d0 /var/lib/containerd/
6929500	/var/lib/containerd/
ethan@ethanm-k8s-upgrade-1:~$ sudo du -d0 -h /var/lib/containerd/
6.7G	/var/lib/containerd/
ethan@ethanm-k8s-upgrade-1:~$ sudo df /
Filesystem     1K-blocks     Used Available Use% Mounted on
/dev/root      203056560 22368024 180672152  12% /
ethan@ethanm-k8s-upgrade-1:~$ sudo df -h /
Filesystem      Size  Used Avail Use% Mounted on
/dev/root       194G   22G  173G  12% /
```

Reply:
looks like an extra .7 gb (725188) in containers and .2.5 gb (979396-725188=254208) in os packages i suppose

Reply:
the original bundle was about a gig. my recommendation for the multiplier is 1x in containerd (extra .7 gb was added) and 2.5x in filesystem (bundle, extracted bundle, os packages at .25 gb)